### PR TITLE
DMXReceive now uses special struct

### PR DESCRIPTION
**Pull request by taylortrimble**
This is a complicated initial diff, so play close attention to what is going on!

You'll notice that some stuff (like waiting for a start address) are not implemented because that's going on in other branches. We won't duplicate the efforts here, but we'll wait for those to materialize.

We'll keep working on it all in parallel! Everyone let me know your thoughts! This is a big change.

### CODE REVIEW

**Comment by taylortrimble** [comment]
Okay, I think this should do the following:
- Accept partial DMX frames as long as we get the data we need
- Utilize a smaller buffer in memory and discard DMX channels we don't need.
- Remove use of global variables for DMX code

But, this is likely to be riddled with little bugs I made. So, @aterlumen and @CrasherCourse especially, scrounge it over and let me know what you think.

**Comment by taylortrimble** [comment]
Yo dawgs, :dog:

Now that you all have a short time for a breather after project 4 is in, take a look at this **_massively complicated diff_** and see if it's something we can merge.

Here we go!

File: DMXlib.c [hunk 1]
```diff
@@ -27,21 +27,20 @@
 #include "DMXlib.h"
 #include <xc.h>
 
-// Constants
-enum {
-    DMXWaitBreak, DMXGotBreak, DMXWaitForStart, DMXWaitForData, DMXDone
+// Types
+
+typedef enum _DMXState {
```
Comment (taylortrimble) [inline DMXlib.c:32]: Note that I changed this from a global variable to a type

File: DMXlib.c [hunk 2]
```diff
@@ -27,21 +27,20 @@
...
+    DMXWaitBreak,
+    DMXGotBreak,
+    DMXWaitForStart,
+    DMXWaitForData,
+    DMXDone
 } DMXState;
 
-// Variables
-int DMXBytesReceived; //16-bit counter
```
Comment (taylortrimble) [inline DMXlib.c:36]: These moved into the function or the new DMX quasi-object

File: DMXlib.h [hunk 3]
```diff
@@ -25,12 +25,15 @@
...
 
 // Defines
-#define DMX_BUFFER_SIZE 512
 #define DMX_START_CODE 0x00
- 
- // Variables
-char DMXBuffer[DMX_BUFFER_SIZE];
+
+// Types
+typedef struct _DMXDevice {
```
Comment (taylortrimble) [inline DMXlib.h:31]: **Important**
Here is the new DMX quasi-object struct. We can set stuff in it to change the behavior of the `DMXReceive` function

**Comment by kevana** [comment]
At first glance the logic looks good to me, but I haven't dug into everything yet.

**Comment by taylortrimble** [comment]
I'm getting trigger happy! Someone stop me! #fingeronmergebutton

**Comment by CrasherCourse** [comment]
The implementation looks fine to me.

However, the pragma section did change  a little after your branch off the master and I don't remember how github will deal with it.

Line 87 has a repeat of ReadDMXStartChannel(void);  I'm not sure if that is to declare first then define later or not.

Also, can we use half words in c?  It might cut down on compare time since 512 needs only 2 bytes instead of 4.

pound holdyourhorses /pound

edit: half words == short 
#justremembered  #poundsarefun

**Comment by taylortrimble** [comment]
Haha, thanks @CrasherCourse!

Yes, git should take care of nice pretty merging for us unless there is a conflict (which it appears I will have to deal with some time) and I was declaring `ReadDMXStartChannel`. We'll have to do that for the rest of them sometime!

I'll look into the `short` type! That's a really good suggestion.

**Comment by taylortrimble** [comment]
**Section 5.4.2 Integer Data Types** of the [xc8 compiler guide](http://www.microchip.com/mplabxc8guide) tells us that `int` and `short int` are both 16 bits already, so we should be good on fast comparisons!

**Comment by CrasherCourse** [comment]
Cool!  Then, there's nothing obvious that I can tell is wrong.

However, I would remove the now defunct CCP2MX pragma that you commented out.

Strangely, the generic integer Wikipedia entry regards int as 4 bytes and it says newer 32+ bit OS's regard int as a full words in C/C++.
But, both the xc8 and c18 compiler use int as shorts so no worries on that.

**Comment by taylortrimble** [comment]
All right! I merged in `master`, and I think this is ready to go. Since:
- No one hates this refactor so far
- It compiles with a smile on its face
- It doesn't seem to have shit logic

I'll probably merge this tomorrow if nothing else comes up!
#merginga23dayoldpullrequestyay :smile:

**Comment by kevana** [comment]
Thumbs up

**Comment by taylortrimble** [comment]
Cool, why wait then? :shipit: Thanks all!

**Pull request merged**